File: code/cpp/q72_edit_distance.cpp
/*
 *  Given two words word1 and word2, find the minimum number of operations required to convert word1 to word2.
 *  You have the following 3 operations permitted on a word:
 *    Insert a character
 *    Delete a character
 *    Replace a character
 */

#include <utility>
#include <vector>
#include <string>
#include <iostream>
#include <unordered_map>

using namespace std;


int getNextSmallestBit(int x) {
    return x ^ (x-1);
}


// delete the bit located at bitIndex (little endian)
int deleteBit(int x, int bitIndex) {
    return x ^ (1 << bitIndex);
}


class Solution {

private:

    typedef vector<int> vi;
    pair<vi, vi> match_letters(string word1, string word2) {
        // these 2 vectors should store integer representations of bit maps
        vector<int> v1, v2;

        char c1('?'), c2('?');
        int wordSize1 = word1.length(), wordSize2 = word2.length(),
            jStartIndex = 0;

        int indexMap1(0), indexMap2(0);
        for (int i = 0; i < wordSize1; ++i) {
            indexMap1 = 0; indexMap2 = 0;

            c1 = word1[i];
            for (int j = jStartIndex; j < wordSize2; ++j) {
                c2 = word2[i];
                if (c1 == c2) {
                    indexMap1 &= (1 << i);
                    indexMap2 &= (1 << j);
                    // v1.push_back(i); v2.push_back(j);
                    // jStartIndex = j + 1;
                    // break;
                }
            }

            if (indexMap1 != 0) {
                v1.push_back(indexMap1); v2.push_back(indexMap2);
            }
        }

        return make_pair(v1, v2);
    }

    // TODO
    int minDistanceForDisjointWords(const string& word1,
                                    const string& word2,
                                    int startIndex1, int endIndex1,
                                    int startIndex2, int endIndex2) {

        int word_size1 = endIndex1 - startIndex1,
            word_size2 = endIndex2 - startIndex2;

        // if word_size1 >
        return 0;
    }

public:
    int minDistance(string word1, string word2) {

    }
};


void print_vector(vector<int> vect) {
    for (int x : vect) {
        cout << x << ' ';
    }
    cout << '\n';
}


int main() {
    Solution sol;
    cout << "Min distance for 'horse', 'ros' "
         << sol.minDistance("horse", "ros") << endl;

    return 0;
}

File: algo_problemset/python/q72_edit_distance.py
"""

72. Edit Distance
Hard
"""

import itertools


class Solution:

    def minDistance(self, word1: str, word2: str) -> int:

        n, m = len(word1), len(word2)
        min_distance = [
            [0 for j in range(m+1)]
            for i in range(n+1)]
        for i in range(n+1): min_distance[i][0] = i
        for j in range(m+1): min_distance[0][j] = j

        for i, j in itertools.product(range(1, n+1), range(1, m+1)):
            if word1[i-1] == word2[j-1]:
                min_distance[i][j] = min_distance[i-1][j-1]

            else:
                min_distance[i][j] = 1 + min(
                    min_distance[i-1][j-1],
                    min_distance[i][j-1],
                    min_distance[i-1][j])

        return min_distance[n][m]


if __name__ == "__main__":
    sol = Solution()

    sol.minDistance("", "a")
